### ISSUE TITLE
Running into issues w/ the git extension

### ISSUE THREAD

**Issue by polarizeme** [Lab41/sunny-side-up#88]
Hiya.

I'll start this off by saying we're not using Docker for this project and we *are* using Jupyterhub so it's very possible those differences are the root of all this. Oh, and I *do* have `gitpython` installed w/ `pip`. That said...

So I'm trying to implement the use of git-commit-push on a jupyterhub server and running into some issues I'm hoping you can comment on.

I was running into an issue on line 24 of git-commit-push.js where, upon writing a commit message and hitting the "commit and push" button, I'm met with the following message in the dev console:
`github-commit-push.js?v=20160425184637:24`
`Uncaught TypeError: Cannot read property '1' of null`

So taking a look at line 24 of the .js file, I see:
`var filepath = window.location.pathname.match(re)[1];`

And the regex above that in line 23 reads:
`var re = /^\/notebooks(.*?)$/;`

Sure enough, running `window.location.pathname.match(/^\/notebooks(.*?)$/)` returns "null."

Since running `window.location.pathname` returns the proper path to the notebook ("/user/[username]/notebooks/[filename].ipynb"), I assumed something must be up with the regex. Again, sure enough, using
`window.location.pathname.match(/\/notebooks\/(.*)$/)[1]`
returns "[filename].ipynb" as intended.

Unfortunately, solving that resulted in the following error on a new attempt to commit:
`github-commit-push.js?v=20160426122528:70`
`PUT https://[server]/git/commit 405 (Method Not Allowed)`
and this is where I'm stumped. I'm not sure what/why this is trying to do with that url on my jupyter[hub] server. My guess is that this is simply relevant to your use of Docker and it's boned on my end because we're not using it, so there are differences in where AJAX calls are meant to be sent.

I know next to nothing about javascript or ajax, let alone your docker environments, so I'm hoping you can chime in with your thoughts on what could be going on.

Sorry this is so long-winded. Ha. Cheers!

**Comment by pcallier** [comment]
Unfortunately the original maintainer of the git extension is unavailable at the moment, so none of us is very well-equipped to diagnose your issue. Our development of the extension may go on, though-- @agude any chance your work is relevant/close to release?

**Comment by agude** [comment]
My work (which should go public at some point in the near future) strips out the github interface; I was mostly interested in committing to local repos. It looks like the bug reported here is with that "Push to Github" step. Working in support for github is something that we'd like to do for my continuation of this plugin, but have not yet.

My best guess at the issue @polarizeme is running into is a change in the API, but that's only my guess from reading the report and not doing any deeper digging. :-/

**Comment by pcallier** [comment]
Yeah that makes sense, though I'd be surprised if GitHub shook up their API so much that even the commit endpoint changed. What may have changed is (a) the method (PUT seems like a weird one; I'd say check the GH API docs) or (b) the authentication part, in which case a more thorough rewrite would basically be unavoidable. I like the workflow in the git extension (where the user takes care of pushing to GitHub), but I can see the appeal of the direct github integration. I will try to replicate your reported behavior at some point in the future, but if you have time to run it down, it would definitely help!

**Comment by pcallier** [comment]
cf. https://github.com/travis-ci/travis.rb/issues/367

**Comment by polarizeme** [comment]
Interesting... The user is able to auth with github creds to get into Jupyterhub (that's the only way we auth for access to it) so I'm curious as to why it would fail for that user on a commit. @pcallier I've only read about 6 words on that travis link, so I'll dig deeper into that and see what I can find that's relevant/helpful. 

I've been retasked to some other stuff for the moment, so it may be a bit before I respond with any updates.

Thanks guys!
/cc @agude

**Comment by agude** [comment]
We'll be here when you return @polarizeme; hopefully with a public repo of the plugin!

**Comment by anaderi** [comment]
@agude are there any chance that news on 'public repo' will appear soon? )

**Comment by nvankayalapati** [comment]
I ran into similar issues with using this extension with Jupyterhub. I got the 
`Uncaught TypeError: Cannot read property '1' of null` which I solved by replacing 
`var filepath = window.location.pathname.match(re)[1];`
with
`var filepath = window.location.pathname.split('notebooks')[1]+'notebooks/';`
The above is very specific to my setup. You may have to do something similar to meet your needs.

After I fixed the above error I got the 
`PUT https://[server]/git/commit 405 (Method Not Allowed)`

The reason for this is because the Javascript code is sending the PUT request to a URL that doesn't have a handler. In the associated Python handler we have
`webapp.add_handlers(".*$", [
        (ujoin(base_url, r"/git/commit"), GitCommitHandler,
            {'log': nbapp.log}),
    ])`

You need to first look at what the `base_url` is. For this you can place this line
`nbapp.log.info(nbapp.web_app.settings['base_url'])`
at the beginning of the `load_jupyter_server_extension` function and look in the logs after you start the Jupyterhub and a single user server.

Once you have this, make sure that your Javascript code sends the PUT request to this correct URL by changing the url field in 
`var settings = {
                    url : base_url+'git/commit',
                    processData : false,
                    headers: {'X-XSRFToken': _xsrf_cookie},
                    type : "PUT",
                    dataType: "json",
                    data: JSON.stringify(payload),
                    contentType: 'application/json',
                    success: success_function,
                    error: error_function
                };`
to include the `base_url`. In my case this turned out to be `var base_url = window.location.pathname.split('notebooks')[0];`. This might be a bit different for you.
Note that I also added an xsrf_cookie because I was having issues with that as well. I generate the `_xsrf_cookie` variable using this code:

`function getCookie(name) {
                var r = document.cookie.match("\\b" + name + "=([^;]*)\\b");
                return r ? r[1] : undefined;
            }
`
`var _xsrf_cookie = getCookie('_xsrf');`

**Comment by sachin235** [comment]
Hi @polarizeme 
Did you manage to find any workaround for this?
I am also trying it for JupyterHub, and stuck at the same issue!

@agude Is there any update on public repo of the plugin which might help to use the plugin on JupyterHub as well?

**Comment by polarizeme** [comment]
@sachin235 sadly, no. But I'll add context to that by saying "being retasked to other stuff for the moment" ended up leading to me quitting my place of employment, thus rendering my project irrelevant haha. 

I'm a jerk and completely forgot to ever come back to this and update as such. If this is stale enough for @pcallier or @agude to mark as closed if necessary, I totally understand.

Sorry for the can of worms and then dissapear.gif ;]

Cheers!

**Comment by sachin235** [comment]
@polarizeme @pcallier @agude 
I have successfully implemented the extension.
https://github.com/sachin235/JhubGitPush

It lets you commit and push directly from JupyterHub to GitHub.

I would like to thank @nvankayalapati for showing the direction to think in!
I have also updated the detailed installation steps in the Readme.
